### ISSUE TITLE
Cache is used for previous information after cache has been updated in a couple methods.

### ISSUE THREAD

**Issue by shelvacu** [serenity-rs/serenity#589]
In [dispatch.rs:269](https://github.com/serenity-rs/serenity/blob/41ff44ba4e4bdd7aa77bfbfce201b89c7990d9e4/src/client/dispatch.rs#L269) the code currently looks like this:

```rust
        DispatchEvent::Model(Event::ChannelUpdate(mut event)) => {
            update!(event);

            let context = context(data, runner_tx, shard_id);
            let event_handler = Arc::clone(event_handler);

            threadpool.execute(move || {
                feature_cache! {{
                    let before = CACHE.read().channel(event.channel.id());

                    event_handler.channel_update(context, before, event.channel);
                } else {
                    event_handler.channel_update(context, event.channel);
                }}
            });
        },
```

Note that `update!`, which updates the cache, comes before `CACHE.read()`, which means that at best `old` will actually be the same channel, and at worst `old` will be a *newer* channel, although that's very unlikely.

There is a similar issue with [GuildUpdate](https://github.com/serenity-rs/serenity/blob/41ff44ba4e4bdd7aa77bfbfce201b89c7990d9e4/src/client/dispatch.rs#L474).

**Comment by xacrimon** [comment]
Might take a look at this myself. We are going to redesign the cache for 0.7 but we might want to fix this before 0.6 releases.

**Comment by Mishio595** [comment]
I suggest taking a look at how I modified MessageUpdate to properly handle this. There are different guarantees with how guilds and channels are cached though, so it might not translate perfectly. https://github.com/serenity-rs/serenity/blob/v0.6.x/src/client/dispatch.rs#L645-L657 and https://github.com/serenity-rs/serenity/blob/v0.6.x/src/model/event.rs#L866-L910 for the CacheUpdate impl.
